tests/test_categorical_model.py

```diff
@@ -345,6 +355,41 @@ def test_matching(
                         (matches[src][tar][expected_label][1] - expected_cost)
                         <= TEST_EPSILON)
 
+    @staticmethod
+    def test_prune(
+            nb_features: int,
+            relation: RelationModel,
+            label_universe: Mapping[int, Tsor],
+            scoring: ScoringModel, algebra: Algebra,
+            arrow: CompositeArrow[int, Tsor],
+            nb_to_prune: int) -> None:
+        """
+            Test pruning operation.
+            If nb_to_prune is a positive integer, tries to prune nb_to_prune
+            relations
+            If nb_to_prune is a negative integer, tries to keep at most
+            nb_to_prune relations + 1
+        """
+        # create cache with one composite arrow
+        cache = TestRelationCache.get_cache(
+            relation, label_universe,
+            scoring, algebra, nb_features, arrow)
+
+        # target number of relations to keep
+        nb_to_keep = (
```

Comment:
I haven't read the code, looking at the test cases first so it might stupid: is it normal nb_to_keep can be negative?

Case: nb_to_prune = -2 => nb_to_keep = nb_to_prune + 1 = -1

Reply:
nb_to_prune = -x is a shorthand to say "I want to keep at most x - 1 relations" (similar to how list indexing works).
I admit it is a bit clunky (although it's explained in the comments of TestRelationCache.test_prune)

Reply:
I don't really see how this is enforced when nb_to_prune is negative. What if the loop returns None in the first step, you may not prune anything, correct?

Reply:
If the loop returns None on the first iteration, this means that no relation can be removed. So yes in this case nothing is pruned.

But this has nothing to do with nb_to_prune being negative. As far as the RelationCache is concerned, it only sees a number of relations to keep which is expected to be positive, the handling of a negative value is done in the test file.

Reply:
Seen offline :-)